### DISCUSSION TITLE
Differences between starlette.testclient.TestClient and httpx.AsyncClient

### DISCUSSION THREAD

**Discussion by antoni-szych-rtbhouse** [encode/starlette#1877] [Q&A] [answered]
Hi,

I was inspired by the release docs, which mention the change in underlying library used in TestClient from `requests` to `httpx`. After reading the [documentation on TestClient](https://www.starlette.io/testclient/), I'm feeling that I don't really understand the purpose of TestClient. We've been using `httpx.AsyncClient` from the moment we started using Starlette in my team, also the documentation [mentions using plain `httpx.AsyncClient`](https://www.starlette.io/testclient/#asynchronous-tests) instead of `starlette.testclient.TestClient`. As `httpx.AsyncClient` works perfectly fine for us, I started wondering: what is the purpose of having `starlette.TestClient` at all? What would be the benefits of using it instead of `httpx.AsyncClient`? And if there aren't any, then why Starlette seem to wrap `httpx.AsyncClient` instead of just telling people in the docs "use `httpx.AsyncClient`, here are some examples for you"?

By looking at the code, it seems like:
- TestClient allows for some backwards-compatibility regarding redirects due to `_choose_redirect_arg`
- TestClient uses `_TestClientTransport` (which seems to be more or less something simmilar to https://github.com/encode/httpx/blob/master/httpx/_transports/asgi.py apart for some additional things related to WebSockets and `Portal`).
- TestClient adds some extra wrapping of ASGI2 apps to behave like ASGI3 apps by `_WrapASGI2`

Maybe the main reason for using starlette.TestClient is that it adds support for websocket connections, which [seems to be lacking out-of-the-box in httpx](https://github.com/encode/httpx/issues/304)? If so, maybe it would be a good idea to write it in the documentation? 
> starlette.TestClient is basically the same as httpx.AsyncClient, however it provides additionally websocket support. If you're using websockets in your application, you need to use starlette.TestClient. Otherwise, it doesn't really matter which one you will use.

Thanks a lot :)

**Comment by euri10** [comment, marked as answer]
I think your analysis is correct, unless there's something I didn;t see in the switch.
I'm too using vanilla httpx for a long time now and to me the only difference with the new TestClient is the websocket part.
I dont intend on switching btw

>  If so, maybe it would be a good idea to write it in the documentation?

yup good call

**Comment by andrew222651** [comment]
A comment: the [docs](https://www.starlette.io/testclient/#asynchronous-tests) mention async issues. I wanted to give a concrete example of how `TestClient` behaves more surprisingly than `AsyncClient`:
```python
"""
run `python -m unittest <this file>`
"""
from unittest import IsolatedAsyncioTestCase
import asyncio

from httpx import AsyncClient
from starlette.responses import HTMLResponse
from starlette.testclient import TestClient


task = None


async def wait_and_print():
    print("sleeping")
    await asyncio.sleep(5)
    print("done")


async def app(scope, receive, send):
    assert scope["type"] == "http"
    response = HTMLResponse("<html><body>Hello, world!</body></html>")
    global task
    task = asyncio.create_task(wait_and_print())
    await response(scope, receive, send)


class TestApp(IsolatedAsyncioTestCase):
    async def test_app(self):
        client = TestClient(app)
        r = client.get("/")
        assert r.status_code == 200
        await asyncio.sleep(10)  # "done" IS NOT printed

    async def test_app_async(self):
        async with AsyncClient(app=app, base_url="http://testserver") as client:
            r = await client.get("/")
            assert r.status_code == 200
            await asyncio.sleep(10)  # "done" IS printed
```
